File: Helper/Bridge/Interface.cs
#pragma warning disable 0, 8981,
// ReSharper disable InconsistentNaming

using Helper;
using Helper.Utility;

namespace Helper.Bridge {

	public static unsafe class Interface {

		#region interface

		[StructLayout(LayoutKind.Sequential)]
		public struct Size {
			public IntegerU64 value;
		}

		[StructLayout(LayoutKind.Sequential)]
		public struct Character {
			public IntegerU8 value;
		}

		[StructLayout(LayoutKind.Sequential)]
		public struct String {
			public Character* data;
			public Size       size;
			public Size       capacity;
		}

		[StructLayout(LayoutKind.Sequential)]
		public struct StringList {
			public String* data;
			public Size    size;
			public Size    capacity;
		}

		[StructLayout(LayoutKind.Sequential)]
		public struct Callback {
			public delegate* <String*, StringList**, StringList**> value;
		}

		// ----------------

		public delegate String* execute (
			Callback**   callback,
			String**     script,
			StringList** argument,
			String**     result
		);

		#endregion

	}

}

File: ShellGUI/lib/bridge/interface.dart
import '/common.dart' as builtin;
import 'dart:ffi' as ffi;

// ----------------

final class Size extends ffi.Struct {
  @ffi.Size()
  external builtin.Integer value;
}

final class Character extends ffi.Struct {
  @ffi.Uint8()
  external builtin.Integer value;
}

final class String extends ffi.Struct {
  external ffi.Pointer<Character> data;
  external Size                   size;
  external Size                   capacity;
}

final class StringList extends ffi.Struct {
  external ffi.Pointer<String> data;
  external Size                size;
  external Size                capacity;
}

final class Callback extends ffi.Struct {
  external ffi.Pointer<ffi.NativeFunction<ffi.Pointer<String> Function(ffi.Pointer<ffi.Pointer<StringList>>, ffi.Pointer<ffi.Pointer<StringList>>)>> value;
}

// ----------------

// ignore: camel_case_types
typedef execute = ffi.Pointer<String> Function(
  ffi.Pointer<ffi.Pointer<Callback>>   callback,
  ffi.Pointer<ffi.Pointer<String>>     script,
  ffi.Pointer<ffi.Pointer<StringList>> argument,
  ffi.Pointer<ffi.Pointer<String>>     result,
);
